### DISCUSSION TITLE
build_tiles consumes huge memory

### DISCUSSION THREAD

**Discussion by clairemibo** [valhalla/valhalla#3288] [Q&A]
Appreciate your patience for reading this question!

We're trying to build tiles from whole world's pbf. We used 24 concurrent threads and 1 GB max cache size for the configuration and we ended up using > 100 GB memory. Our target is to make it under 50 GB but finding it difficult. 

We've also tried 16 threads + 600MB max_cache_size - not helping that much. So we profiled the memory usage and saw that the Enhance and Validation steps are the parts consuming most of the memory. Per enhance step, the "GetDensity" function and "GetGraphTile" are the top 2 memory intensive places.

Are we missing anything here? Any suggestions on how we can make it consumes less than 50 GB? Thanks!

**Comment by kevinkreiser** [comment]
so i run this on a machine with 32 cores and 64gigs of ram. i set the code to use all the cores and i use the default config which is a cache size of 1GB. the tile building never gets OOM killed for me. Are you saying that yours does get OOM killed?

I would be surprised if the enhance step is the largest memory consumer depending on how you are measuring it. During the build stage we memory map 2 or 3 huge files, probably somewhere around 100GB total. This causes the machine to report a virtual memory usage that is as high as the address ranges assigned to those files (so somewhere around 100GB). this doesnt mean that the computer is actually using that much ram, indeed for me, the computer doestn even have that much ram.

can you maybe let us know how you are measuring ram usage? are you using the standard OSM planet or do you have your own custom dataset?

**Reply by gknisely** [reply]
@clairemibo where are you seeing the huge memory consumption?  During the build phase?

**Reply by mandeepsandhu** [reply]
You can look at a process' `RSS` usage (resident set size) to determine how much 'RAM' its consuming:
```shell
$ cat /proc/<pid>/status | grep -i vmrss
```

**Reply by clairemibo** [reply]
@kevinkreiser, thanks for you reply! We have custom dataset based on standard OSM (basically, we're adding some ways/nodes and filtering out some ways/nodes we don't need per our use case).  The size of our_planet.pbf is only 30G while the standard OSM pbf is 59G.

The memory usage we got (>100G) is the usage that the linux cgroup reported for the process.

**Reply by nilsnolde** [reply]
What OSM ID range do you choose for your synthetic data? Even though it wasn’t Valhalla, it was osmium, I ran into that problem before too. Especially if you don’t properly sort the data in the resulting pbf (Valhalla does complain about unsorted data, but not sure if that’s not only regarding the type sorting, as opposed to ID sorting within types).

**Reply by clairemibo** [reply]
@nilsnolde , thanks for checking in! The ID we're choosing is less than max_int32. We never had id issues when running build_tiles. We use osmium to sort our planet.pbf.

**Reply by nilsnolde** [reply]
ok, then that's definitely not an issue.. I'd still be careful though with using a max of int32, OSM is long past 32 bit, so you might get duplicate id's (but maybe you're somehow careful about that?). not entirely where we currently are, but it's safest to start synthetic data at minimum 36 bits (up to 64), most routing engines should be able to deal with that by now.

**Reply by nilsnolde** [reply]
one more word of caution: be careful about altering OSM data (in case you're not opening your database to the general public); simply removing nodes/ways should be fine, adding or editing anything is pretty sure against OSM's ODBL: https://wiki.openstreetmap.org/wiki/Open_Database_License

**Reply by clairemibo** [reply]
@nilsnolde Thanks for the caution! We'll keep that in mind.

**Comment by mandeepsandhu** [comment]
> So we profiled the memory usage and saw that the Enhance and Validation steps are the parts consuming most of the memory. Per enhance step, the "GetDensity" function and "GetGraphTile" are the top 2 memory intensive places.

@gknisely :point_up: looks like its the enhance & validation steps.

**Reply by davelowndes** [reply]
>  i set the code to use all the cores

@kevinkreiser How do you do this? Can you do it with docker image ghcr.io/valhalla/valhalla:3.4.0? I'd appreciate your help as I'm getting oom with a full planet build (#4855). I can double the resources to 32 cores and 64gigs of ram but want to ensure the rest of the configure is correct. 🙏

**Reply by kevinkreiser** [reply]
try master, we havent released in a while and because of that there are lots of things (specifically work that @gknisely did to bring down the ram usage) that could improve your situation
